### PR TITLE
[FLINK-17515][yarn] Move file upload functionality out of the YarnClusterDescriptor

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

**NOTE TO REVIEWERS:** all commits in this PR are going to get squashed into a single one, when merged. I just left them here in case it is easier to follow the changes.

This PR refactors code out of the `YarnClusterDescriptor` and the `yarn.Utils` and into the `YarnApplicationFileUploader`. The main methods transferred are the:

* `Utils.uploadLocalFileToRemote()`
* `uploadAndRegisterFiles()`
* `setupSingleLocalResource()`

## Verifying this change

This change is a trivial rework / code cleanup and it is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes - YARN** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+		// construct classpath, we always want resource directories to go first, we also sort
+		// both resources and archives in order to make classpath deterministic
+		final ArrayList<String> classPaths = new ArrayList<>();
+		resources.stream().sorted().forEach(classPaths::add);
+		archives.stream().sorted().forEach(classPaths::add);
+		return classPaths;
+	}
+
+	static YarnApplicationFileUploader initialize(
```
Comment (aljoscha) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationFileUploader.java:220]: shouldn't this be `create()` or `of()`?

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 2]
```diff
@@ -901,10 +885,6 @@ private ApplicationReport startAppMaster(
 			}
 		}
 
-		final Path yarnFilesDir = getYarnFilesDir(appId);
```
Comment (aljoscha) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:904]: Why don't we need the permission setting anymore?

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 3]
```diff
@@ -1105,160 +1076,6 @@ private ApplicationReport startAppMaster(
...
-				relativePaths.add(new Path(localResourcesDirectory, shipFile.getName()));
-			}
-		}
-
-		final Set<String> archives = new HashSet<>();
-		final Set<String> resources = new HashSet<>();
-		for (int i = 0; i < localPaths.size(); i++) {
-			final Path localPath = localPaths.get(i);
-			final Path relativePath = relativePaths.get(i);
-			if (!isDistJar(relativePath.getName())) {
```
Comment (aljoscha) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1232]: Why don't we need this "dist jar" business anymore?

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 2 above)
Comment (kl0u) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:904]: It is happening in the `YarnApplicationFileUploader.getApplicationDir()`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 3 above)
Comment (kl0u) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1232]: This method has moved entirely to the new class.

Comment (aljoscha) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:1232]: Ah, I know saw that you renamed the method in the moved code. I was looking for the same method name.

**Pull request closed without merging**